File: 2024/day08/solution.py
#!/usr/bin/env python3


class Coord():

  def __init__(self, row: int, col: int):
    self._val = (row, col)

  def __add__(self, other):
    if isinstance(other, Coord):
      return Coord(
        self[0] + other[0],
        self[1] + other[1],
      )
    elif isinstance(other, (int, float)):
      return Coord(
        self[0] + other,
        self[1] + other,
      )
    else:
      raise TypeError('operand must be a coord, int, or float')

  def __mul__(self, other):
    if isinstance(other, Coord):
      return Coord(
        self[0] * other[0],
        self[1] * other[1],
      )
    elif isinstance(other, (int, float)):
      return Coord(
        self[0] * other,
        self[1] * other,
      )
    else:
      raise TypeError('operand must be a coord, int, or float')

  def __sub__(self, other):
    if isinstance(other, Coord):
      return Coord(
        self[0] - other[0],
        self[1] - other[1],
      )
    elif isinstance(other, (int, float)):
      return Coord(
        self[0] - other,
        self[1] - other,
      )
    else:
      raise TypeError('operand must be a coord, int, or float')

  def __hash__(self):
    return self._val.__hash__()

  def __eq__(self, other):
    return self._val.__hash__() == other._val.__hash__()

  def __getitem__(self, ix: int):
    return self._val[ix]

  def __repr__(self):
    return self._val.__repr__()

  def simplify(self):
    abs_vals = [abs(self._val[0]), abs(self._val[1])]
    large, small = (max(abs_vals), min(abs_vals))
    while (remainder := large % small):
      large, small = (max(small, remainder), min(small, remainder))
    gcd = small
    return Coord(
      self._val[0] // gcd,
      self._val[1] // gcd,
    )


def draw_map(rows: int, cols: int, freq_map: dict, freq: str, props: list[str]):
  print()
  for ix_row in range(rows):
    line = ''
    for ix_col in range(cols):
      coord = Coord(ix_row, ix_col)
      if coord in freq_map['antennas']:
        line += freq
      elif any(coord in freq_map[x] for x in props):
        line += '#'
      else:
        line += '.'
    print(line)


with open('input.txt', 'r') as f:
  data = [list(x) for x in f.read().splitlines()]

count_rows = len(data)
count_cols = len(data[0])

freq_maps = {
  frequency: {
    'antennas': {
      Coord(ix_row, ix_col)
      for ix_row in range(count_rows)
      for ix_col in range(count_cols)
      if data[ix_row][ix_col] == frequency
    },
    'antinodes': set(),
    'antinodes_2': set(),
  }
  for frequency in {col
                    for row in data
                    for col in row}
  if frequency != '.'
}

# part 1 antinodes
for freq, freq_map in freq_maps.items():
  for antenna_a in freq_map['antennas']:
    for antenna_b in freq_map['antennas']:
      if antenna_a == antenna_b: continue
      antinode_a = antenna_a - (antenna_b - antenna_a)
      antinode_b = antenna_b - (antenna_a - antenna_b)
      for antinode in (antinode_a, antinode_b):
        if 0 <= antinode[0] < count_rows and 0 <= antinode[1] < count_cols:
          # print(f'{freq=} {antinode=}')
          freq_map['antinodes'].add(antinode)

for freq, freq_map in freq_maps.items():
  draw_map(count_rows, count_cols, freq_map, freq, ['antinodes'])

antinodes = {c for m in freq_maps.values() for c in m['antinodes']}

# print(f'{antinodes=}')
print(f'part 1 {len(antinodes)=}')

# part 2 antinodes
for freq, freq_map in freq_maps.items():
  for antenna_a in freq_map['antennas']:
    for antenna_b in freq_map['antennas']:
      if antenna_a == antenna_b: continue
      for antenna, delta in (
        (antenna_a, (antenna_b - antenna_a).simplify()),
        (antenna_b, (antenna_a - antenna_b).simplify()),
      ):
        multiplier = 1
        while True:
          antinode = antenna + (delta * multiplier)
          if 0 <= antinode[0] < count_rows and 0 <= antinode[1] < count_cols:
            # print(f'{freq=} {antinode=}')
            freq_map['antinodes_2'].add(antinode)
            multiplier += 1
          else:
            break

for freq, freq_map in freq_maps.items():
  draw_map(count_rows, count_cols, freq_map, freq, ['antinodes_2'])

antinodes2 = {c for m in freq_maps.values() for c in m['antinodes_2']}
print(f'part 2 {len(antinodes2)=}')

File: 2024/day08/solution.ts
#!/usr/bin/env -S deno --allow-read

const DEBUG = true;
// const DEBUG = false;

class Coord {
  constructor(public row: number, public col: number) {}
  hash(): number {
    return this.row * 10000000 + this.col;
  }
}

const text = await Deno.readTextFile('input.txt');
if (DEBUG) console.debug(text);

const mapData = text
  .split('\n')
  .filter((line) => line.trim() != '')
  .map((line) => line.split(''));
const countRows = mapData.length;
const countCols = mapData[0].length;
if (DEBUG) console.debug({ mapData, countRows, countCols });

// const antennas: { frequency: string; position: number[] }[] = [];
const frequencies: Record<string, { antennas: Map<number, Coord>; antinodes: Map<number, Coord> }> = {};
for (let ixRow = 0; ixRow < countRows; ixRow++) {
  for (let ixCol = 0; ixCol < countCols; ixCol++) {
    const frequency = mapData[ixRow][ixCol];
    if (frequency == '.') continue;
    const antenna = new Coord(ixRow, ixCol);
    if (frequencies[frequency]) {
      frequencies[frequency].antennas.set(antenna.hash(), antenna);
    } else {
      frequencies[frequency] = {
        antennas: new Map([[antenna.hash(), antenna]]),
        antinodes: new Map(),
      };
    }
  }
}
if (DEBUG) console.debug({ frequencies });

Object.entries(frequencies).forEach(([frequency, props]) => {
  if (DEBUG) console.debug({ frequency, props });
  for (const [hash0, antenna0] of props.antennas) {
    for (const [hash1, antenna1] of props.antennas) {
      if (hash0 == hash1) continue;
      for (const antinode of [
        new Coord(antenna0.row + antenna0.row - antenna1.row, antenna0.col + antenna0.col - antenna1.col),
        new Coord(antenna1.row + antenna1.row - antenna0.row, antenna1.col + antenna1.col - antenna0.col),
      ]) {
        if (antinode.row >= 0 && antinode.row < countRows && antinode.col >= 0 && antinode.col < countCols) {
          props.antinodes.set(antinode.hash(), antinode);
        }
      }
    }
  }
});
if (DEBUG) console.debug({ frequencies });

const allAntinodes = Object.entries(frequencies).reduce((acc, [_, props]) => {
  for (const [hash, antinode] of props.antinodes) {
    acc.set(hash, antinode);
  }
  return acc;
}, new Map<number, Coord>());
if (DEBUG) console.debug({ allAntinodes, countRows, countCols });
console.log('part 1', allAntinodes.size);

Object.entries(frequencies).forEach(([frequency, props]) => {
  if (DEBUG) console.debug({ frequency, props });
  props.antinodes.clear();
  for (const [hash0, antenna0] of props.antennas) {
    for (const [hash1, antenna1] of props.antennas) {
      if (hash0 == hash1) continue;
      const baseOffset = new Coord(antenna1.row - antenna0.row, antenna1.col - antenna0.col);
      for (let i = countRows * -1; i < countRows; i++) {
        const antinode = new Coord(antenna0.row + baseOffset.row * i, antenna0.col + baseOffset.col * i);
        if (antinode.row >= 0 && antinode.row < countRows && antinode.col >= 0 && antinode.col < countCols) {
          props.antinodes.set(antinode.hash(), antinode);
        }
      }
    }
  }
});
if (DEBUG) console.debug({ frequencies });

const allAntinodes2 = Object.entries(frequencies).reduce((acc, [_, props]) => {
  for (const [hash, antinode] of props.antinodes) {
    acc.set(hash, antinode);
  }
  return acc;
}, new Map<number, Coord>());
if (DEBUG) console.debug({ allAntinodes2, countRows, countCols });
console.log('part 2', allAntinodes2.size);
